### ISSUE TITLE
Propose syntax for container entities

### ISSUE THREAD

**Issue by gimer** [nemtech/catbuffer#2]
With upcoming schmea changes we'll have aggregate transaction.
Currently `array()` is supposed to handle `number  of elements.
I.e:
```
	# number of modifications
	modificationsCount = uint8

	# attached modifications
	modifications = array(CosignatoryModification, modificationsCount)
```

modificationsCount is a number of CosignatoryModification structs

for aggregate transaction however we need to be able to:
a) bind payloadSize to BYTE-based size of attached transactions
b) express difference *somehow*
Currently this is done via:
```
	transactions = array(byte, payloadSize)
	cosignatures = array(byte, size - payloadSize)
```

transactions is actually container for transactions, where each is actually `SizePrefixedEntity` [schemas/entity.cats](https://github.com/nemtech/catbuffer/blob/master/schemas/entity.cats)

cosignatures is actually array of `Cosignature` structs, but size needs to be defined (bound) to difference between tx size and payloadSize.

so:
 1. we rather should not use array() for that purpose, (some new keyword is needed along with defined meaning)
 2. not sure if binding to difference of two sizes makes sense, generators would probably benefit from some nicer syntax, maybe something like:
```
transactions = container(Transactions)
bind(payloadSize, transactions.size)

cosignatures = container(Cosignature)
bind(size, cosignatures.size + payloadSize)
```

note: there is a slight difference between two in sample above:
 * Transaction is a SizePrefixedEntity, while
 * Cosignature is a fixed-size struct

so maybe common syntax for both of them is also bad idea...

**Comment by Vektrat** [comment]
i'm not sure i understand the problem correctly, but...

i believe the transactions collection is an array and should stay so, it is true that up until now arrays were composed of fixed size elements, so wouldn't it make sense to have the same notation as before, and be able to tell if the contained elements are of fixed size, or variable? we could introduce a new static value to the schema, or we could just check if the schema has a `size` field if this is good enough - if catbuffer detects variable size, then it would calcualte size via .getSize() of each sub-element

i don't understand the problem with the cosignatures array... isn't it like the previous situations? an array of a determinate number of elements (i'd expect we'd have `transactionsCount` and `cosignaturesCount`) of fixed size (`Cosignature`)

so...
```
transactionsCount = uint8
cosignaturesCount = uint8

# treated differently because catbuffer will somehow know (TBD) Transaction is of variable size
transactions = array(Transaction, transactionsCount)

# treated like an array as before
cosignatures = array(Cosignature, cosignaturesCount)
```

**Comment by gimer** [comment]
the problem is that we need to have syntax, that matches already existing models...

1. in case of blocks:
  a. you have .Size field, and you need to iterate through block to count elements (that is because if you had .Count, you couldn't trust that field anyway)
  b. you have dependency `.Size = "header size" + summed txes size` that has to be expressed *somehow*, (so that catbuffer-based generators are able to create serialized model with .Size field properly set)
2. In case of aggregate transactions, similar to block, but
  a. dependency 1 `.PayloadSize = summed txes size`
  b. dependency 2 `.Size = "header size" + .PayloadSize + cosignatures size`
 

so, say i have bunch of generated builders, I want to be able to do:
```
transferBuilder.setRecipient()
...
tx = transferBuilder.buildEmbedded()


aggregateBuilder.addTransaction(tx)
aggregateBuilder.addCosignature(cosig1)
aggregateBuilder.addCosignature(cosig2)
finalTransaction = aggregateBuilder.build()
```

hope that helps.

**Comment by Jaguar0625** [comment]
Fixed in 199c607.
